istio/src/main/scala/io/buoyant/k8s/istio/mixer/MixerClient.scala

```diff
@@ -0,0 +1,118 @@
+import com.twitter.finagle.buoyant.h2
+import com.twitter.logging.Logger
+import com.twitter.util.{Future, Return, Throw}
+import io.buoyant.grpc.runtime.GrpcStatus._
+import io.buoyant.grpc.runtime.{GrpcStatus, Stream}
+import io.buoyant.k8s.istio._
+import istio.mixer.v1.{CheckResponse, Mixer}
+
+case class MixerCheckStatus(grpcStatus: GrpcStatus) {
+  def success = grpcStatus match {
+    case _: GrpcStatus.Ok => true
+    case _ => false
+  }
+
+  def code = grpcStatus.code
+
+  def reason = s"${grpcStatus.getClass.getSimpleName} - ${grpcStatus.code} - ${grpcStatus.message}"
+
+  /**
+   * Converts between Mixer's status codes and HTTP status codes, as per [[https://github.com/istio/proxy/blob/7b85ad9cdfa460e928dd24b6c779b1b0e82fff4f/src/envoy/mixer/http_filter.cc#L66 Istio's default proxy]]
+   */
+  def httpCode = grpcStatus match {
+    case _: Ok => 200
+    case _: Canceled => 499
+    case _: Unknown => 500
+    case _: InvalidArgument => 400
+    case _: DeadlineExceeded => 504
+    case _: NotFound => 404
+    case _: AlreadyExists => 409
+    case _: PermissionDenied => 403
+    case _: ResourceExhausted => 429
+    case _: FailedPrecondition => 400
+    case _: Aborted => 409
+    case _: OutOfRange => 400
+    case _: Unimplemented => 501
+    case _: Internal => 500
+    case _: Unavailable => 503
+    case _: DataLoss => 500
+    case _: Unauthenticated => 401
+    case _: Other => 500
+  }
+}
+
+class MixerClient(client: Mixer) {
+  private[this] val log = Logger(this.getClass.getSimpleName)
+  private[this] val codeUsedWhenUnknownStatus = Unknown("NO RESPONSE FROM SERVER")
+
+  /**
+   * Sends a report to Istio' Mixer, using the [[https://istio.io/docs/reference/api/mixer/mixer-service.html#report Report API]]
+   *
+   * minimum set of attributes to generate the following metrics in mixer/prometheus:
+   * - request_count
+   * - request_duration_bucket
+   * - request_duration_count
+   * - request_duration_sum
+   * *
+   * example metrics exposed by mixer/prometheus:
+   * request_count{method="/productpage",response_code="200",service="productpage",source="unknown",target="productpage.default.svc.cluster.local",version="v1"} 327
+   * request_count{method="/reviews",response_code="500",service="reviews",source="productpage",target="reviews.default.svc.cluster.local",version="v1"} 1
+   */
+  def report(
+    responseCode: ResponseCodeIstioAttribute,
+    requestPath: RequestPathIstioAttribute,
+    targetService: TargetServiceIstioAttribute,
+    sourceLabel: SourceLabelIstioAttribute,
+    targetLabel: TargetLabelsIstioAttribute,
+    duration: ResponseDurationIstioAttribute
+  ): Future[Unit] = {
+
+    val reportRequest = MixerApiRequests.mkReportRequest(
+      responseCode,
+      requestPath,
+      targetService,
+      sourceLabel,
+      targetLabel,
+      duration
+    )
+    log.trace("MixerClient.report: %s", reportRequest)
+    client.report(Stream.value(reportRequest))
+    Future.Done
```

Comment:
Should this return a finished `Future`, or return a future that's satisfied when the report stream finishes?

Reply:
It's definitely fire-and-forget, and because nobody cares about when it's completed I decided to just return a constant. On the return type, I generally like methods from a single type to be symmetric, i.e. given a set `A` it always map from `A` to `S[A]`, but no strong preference.

Reply:
Hmmm... I don't understand your comment about symmetric return types.  What would be the difference between having the return type be `Unit` instead of `Future[Unit]`?

Reply:
I mean that I would expect all functions in a given type/namespace to return values of the same container, or at least the same _kind_. In this particular case, it means that I would expect the published interface of `MixerClient` to be either this:

```scala
class MixerClient(client: Mixer) {
  def report(/* ... */): Future[Unit]
  def checkPreconditions(/* ... */): Future[MixerCheckStatus]
}
```

Or:
```scala
class MixerClient(client: Mixer) {
  def report(/* ... */): Unit
  def checkPreconditions(/* ... */): MixerCheckStatus
}
```

But **not**:
```scala
class MixerClient(client: Mixer) {
  def report(/* ... */): Unit
  def checkPreconditions(/* ... */): Future[MixerCheckStatus]
}
```

But, again, this is just a preference, I'm happy to make it `Unit` if it makes it more aligned with our internal philosophy.

Reply:
Ah, I see.  I think `Unit` is more aligned with my personal philosophy anyway.  (since this method is fire-and-forget)